jerry-core/ecma/base/ecma-helpers-errol.c

```diff
@@ -0,0 +1,240 @@
+ */
+
+/**
+ * Floating point format definitions
+ */
+#define FP_MAX DOUBLE_MAX
+#define FP_MIN DOUBLE_MIN
+#define fpnext(val) (nextafter ((val), INFINITY))
+#define fpprev(val) (nextafter ((val), -INFINITY))
+
+#define ERROL0_EPSILON 0.0000001
+#define ERROL1_EPSILON 8.77e-15
+
+/**
+ * High-precision data structure.
+ */
+struct hp_t
+{
+  ecma_number_t val; /**< value */
+  ecma_number_t off; /**< offset */
+};
+
+/**
+ * Normalize the number by factoring in the error.
+ */
+static inline void __attr_always_inline___
+hp_normalize (struct hp_t *hp) /**< [in, out] float pair */
+{
+  ecma_number_t val = hp->val;
+
+  hp->val += hp->off;
+  hp->off += val - hp->val;
+} /* hp_normalize */
+
+/**
+ * Multiply the high-precision number by ten.
+ */
+static inline void __attr_always_inline___
+hp_mul10 (struct hp_t *hp) /**< [in, out] high-precision number */
+{
+  ecma_number_t off, val = hp->val;
+
+  hp->val *= 10.0;
+  hp->off *= 10.0;
+
+  off = hp->val;
+  off -= val * 8.0;
+  off -= val * 2.0;
+
+  hp->off -= off;
+
+  hp_normalize (hp);
+} /* hp_mul10 */
+
+/**
+ * Divide the high-precision number by ten.
+ */
+static void
+hp_div10 (struct hp_t *hp) /**< [in, out] high-precision number */
+{
+  ecma_number_t val = hp->val;
+
+  hp->val /= 10.0;
+  hp->off /= 10.0;
+
+  val -= hp->val * 8.0;
+  val -= hp->val * 2.0;
+
+  hp->off += val / 10.0;
+
+  hp_normalize (hp);
+} /* hp_div10 */
+
+/**
+ * Errol0 double to ASCII conversion, guaranteed correct but possibly not optimal.
+ *
+ * @return the exponent
+ */
+inline int32_t __attr_always_inline___
+ecma_errol0_dtoa (ecma_number_t val, /**< ecma number */
+                  uint64_t *digits_p, /**< [out] digits */
+                  int32_t *num_of_digits_p) /**< [out] number of digits */
+{
+  double ten;
+  int32_t exp;
+  struct hp_t mid, inhi, inlo;
+  uint64_t digits = 0u;
+  int32_t num_of_digits = 0;
+  ten = 1.0;
+  exp = 1;
+
+  /* normalize the midpoint */
+
+  mid.val = val;
+  mid.off = 0.0;
+
+  while (((mid.val > 10.0) || ((mid.val == 10.0) && (mid.off >= 0.0))) && (exp < 308))
+  {
+    exp++;
+    hp_div10 (&mid);
+    ten /= 10.0;
+  }
+
+  while (((mid.val < 1.0) || ((mid.val == 1.0) && (mid.off < 0.0))) && (exp > -307))
+  {
+    exp--;
+    hp_mul10 (&mid);
+    ten *= 10.0;
+  }
+
+  inhi.val = mid.val;
+  inhi.off = mid.off + (fpnext (val) - val) * ten / (2.0 + ERROL0_EPSILON);
+  inlo.val = mid.val;
+  inlo.off = mid.off + (fpprev (val) - val) * ten / (2.0 + ERROL0_EPSILON);
+
+  hp_normalize (&inhi);
+  hp_normalize (&inlo);
+
+  /* normalized boundaries */
+
+  while (inhi.val > 10.0 || (inhi.val == 10.0 && (inhi.off >= 0.0)))
+  {
+    exp++;
+    hp_div10 (&inhi);
+    hp_div10 (&inlo);
+  }
+
+  while (inhi.val < 1.0 || (inhi.val == 1.0 && (inhi.off < 0.0)))
+  {
+    exp--;
+    hp_mul10 (&inhi);
+    hp_mul10 (&inlo);
+  }
+
+  /* digit generation */
+
+  while (inhi.val != 0.0 || inhi.off != 0.0)
+  {
+    uint8_t ldig, hdig = 0;
```

Comment:
initialization with 0 is unnecessary, both `ldig` and `hdig` will be assigned just some lines below; better move their declaration to their initialization